src/index.ts

```diff
@@ -0,0 +1,408 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Utility functions for handling suggestions.
+ */
+import * as Blockly from 'blockly/core';
+import IMachineLearningModel from './IMachineLearningModel';
+
+/** Map from workspaces to BlockSuggestor objects. */
+export const suggestorLookup = new WeakMap<Blockly.Workspace, BlockSuggestor>();
+
+/**
+ * Class that tracks all blocks created in a workspace and suggests future
+ * blocks to use.
+ */
+export class BlockSuggestor {
+  /**
+   * Machine learning model for block suggestions
+   */
+  private model: IMachineLearningModel | null = null;
+  /**
+   * Saves the full JSON data for each block type the first time it's used.
+   * This helps store what initial configuration / sub-blocks each block type
+   * would be expected to have.
+   */
+  private defaultJsonForBlockLookup: Record<
+    string,
+    Blockly.utils.toolbox.BlockInfo
+  > = {};
+  /**
+   * List of recently used block types
+   */
+  private recentlyUsedBlocks: string[] = [];
+  /**
+   * Checks if the workspace is finished loading, to avoid taking action on
+   * all the BLOCK_CREATE events during workspace loading.
+   */
+  /* eslint-disable-next-line @typescript-eslint/explicit-member-accessibility */
+  public workspaceHasFinishedLoading = false;
+
+  /**
+   * Config parameter which sets the size of the toolbox categories.
+   */
+  private numBlocksPerCategory: number;
+
+  /**
+   * The Blockly workspace instance associated with this BlockSuggestor.
+   */
+  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+  public workspace: Blockly.WorkspaceSvg | null = null;
+  /**
+   * Optional input description element or string
+   */
+  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+  public inputSource: HTMLInputElement | string | null = null;
+  /**
+   * Constructs a BlockSuggestor object.
+   *
+   * @param numBlocksPerCategory the size of each toolbox category
+   * @param workspace new workspace
```

Comment:
This `workspace` need not be new as the user may pass a reference to one as an argument to `init` instead of an `HTMLElement`. You want a comment that describes its relation with respect to the `BlockSuggestor` object.